Source/DafnyCore/AST/Grammar/ParseErrors.cs

```diff
@@ -407,18 +510,562 @@ declared identifier.
+You must use parentheses to show how they are grouped. Write `p && q || r` as either
+`(p && q) || r` or `p && (q || r)`.
+");// TODO - would be nice to have code actions for the alternatives, but that requires passing multiple ranges
+
+    Add(ErrorId.p_invalid_equal_chaining,
+    @"
+[Chained operations](../DafnyRef/DafnyRef#sec-basic-types)
+are a sequence of binary operations without parentheses: _a op b op c op d op e_.
+But there are limitations on which operators can be in one chain together.
+
+In particular, the relational operators `in` and `!in` may not be part of a chain.
+Use parentheses as necessary to group the operations.
+");
+
+    Add(ErrorId.p_invalid_notequal_chaining,
+    @"
+[Chained operations](../DafnyRef/DafnyRef#sec-basic-types)
+are a sequence of binary operations without parentheses: _a op b op c op d op e_.
+But there are limitations on which operators can be in one chain together.
+
+In particular for this error message, one cannot have chains that include more than one `!=` operator.
+");
+
+    Add(ErrorId.p_invalid_descending_chaining,
+    @"
+[Chained operations](../DafnyRef/DafnyRef#sec-basic-types)
+are a sequence of binary operations without parentheses: _a op b op c op d op e_.
+But there are limitations on which operators can be in one chain together.
+
+In particular for this error message, one cannot have chains that include both
+less-than operations (either `<` or `<=`) and greater-than operations (either `>` or `>=`).
+");
+
+    Add(ErrorId.p_invalid_ascending_chaining,
+    @"
+[Chained operations](../DafnyRef/DafnyRef#sec-basic-types)
+are a sequence of binary operations without parentheses: _a op b op c op d op e_.
+But there are limitations on which operators can be in one chain together.
+
+In particular for this error message, one cannot have chains that include both
+less-than operations (either `<` or `<=`) and greater-than operations (either `>` or `>=`).
+");
+
+    Add(ErrorId.p_invalid_disjoint_chaining,
+    @"
+[Chained operations](../DafnyRef/DafnyRef#sec-basic-types)
+are a sequence of binary operations without parentheses: _a op b op c op d op e_.
+But there are limitations on which operators can be in one chain together.
+
+In particular for this error message, a disjoint operator (`!!`) can appear in a chain
+only if all the operators in the chain are disjoint operators.
+
+As described [here](../DafnyRef/DafnyRef#sec-collection-types),
+`a !! b !! c !! d` means that `a`, `b`, `c`, and `d` are all mutually disjoint
+(which is a different rewriting of the chain than for other operations).
+");
+
+    Add(ErrorId.p_operator_does_not_chain,
+    @"
+The operators `in` and `!in` are relational operators, but they may not occur in a chain.
+Use parentheses if necessary. Such expressions are usually not type-correct in any case.
+");
+
+    Add(ErrorId.p_bang_not_a_relational_op,
+    @"
+const s : set<int>
+const r := s ! s
+", Replace("!="));
```

Comment:
Can you add also "!!" ? In relational operators, it's another one that says "disjoint" and it's also possible the user meant it.

Reply:
Done